src/modules/ekf2/EKF/aid_sources/external_vision/ev_vel_control.cpp

```diff
@@ -256,92 +144,64 @@ void Ekf::controlEvVelFusion(const imuSample &imu_sample, const extVisionSample
 	}
 }
 
-bool Ekf::fuseEvVelocity(estimator_aid_source3d_s &aid_src, const extVisionSample &ev_sample)
+void Ekf::stopEvVelFusion()
 {
-	if (ev_sample.vel_frame == VelocityFrame::BODY_FRAME_FRD) {
-
-		VectorState H;
-		estimator_aid_source1d_s current_aid_src;
-		const auto state_vector = _state.vector();
-
-		for (uint8_t index = 0; index <= 2; index++) {
-			current_aid_src.timestamp_sample = aid_src.timestamp_sample;
+	if (_control_status.flags.ev_vel) {
+		_control_status.flags.ev_vel = false;
+	}
+}
 
-			if (index == 0) {
-				sym::ComputeEvBodyVelHx(state_vector, &H);
+void Ekf::fuseLocalFrameVelocity(estimator_aid_source3d_s &aid_src, const uint64_t &timestamp,
+				 const Vector3f &measurement, const Vector3f &measurement_var, const float &innovation_gate)
+{
+	updateAidSourceStatus(aid_src,
+			      timestamp,				// sample timestamp
+			      measurement,				// observation
+			      measurement_var,				// observation variance
+			      _state.vel - measurement,			// innovation
+			      getVelocityVariance() + measurement_var,	// innovation variance
+			      innovation_gate);				// innovation gate
+	fuseVelocity(aid_src);
+}
 
-			} else if (index == 1) {
-				sym::ComputeEvBodyVelHy(state_vector, &H);
+void Ekf::fuseBodyFrameVelocity(estimator_aid_source3d_s &aid_src, const uint64_t &timestamp,
+				const Vector3f &measurement, const Vector3f &measurement_var, const float &innovation_gate)
+{
+	const auto state_vector = _state.vector();
 
-			} else {
-				sym::ComputeEvBodyVelHz(state_vector, &H);
-			}
+	matrix::Vector<VectorState, 3> H;
 
-			const float innov_var = (H.T() * P * H)(0, 0) + aid_src.observation_variance[index];
-			const float innov = (_R_to_earth.transpose() * _state.vel - Vector3f(aid_src.observation))(index, 0);
+	sym::ComputeBodyVelH(state_vector, &H(0), &H(1), &H(2));
 
-			updateAidSourceStatus(current_aid_src,
-					      ev_sample.time_us,				// sample timestamp
-					      aid_src.observation[index],			// observation
-					      aid_src.observation_variance[index],		// observation variance
-					      innov,						// innovation
-					      innov_var,    					// innovation variance
-					      math::max(_params.ev_vel_innov_gate, 1.f));	// innovation gate
+	Vector3f innov = _R_to_earth.transpose() * _state.vel - measurement;
+	Vector3f innov_var;
 
-			if (!current_aid_src.innovation_rejected) {
-				fuseBodyVelocity(current_aid_src, current_aid_src.innovation_variance, H);
+	for (uint8_t index = 0; index <= 2; index++) {
+		innov_var(index) = (H(index).T() * P * H(index))(0, 0) + measurement_var(index);
```

Comment:
Computing an innovation variance using raw matrix multiplications (H * P * H') involves a lot of operations, did you compare it with the auto-generated one?

Reply:
I compared it:
calculate HPH inline:
Average time: 2.54625e-05s
  // Total ops: 138
  
use symforce autogenerated:
Average time: 2.14646e-05s
  // Total ops: 371
  
Regarding the number of operations I would prefer saving the flash